File: src/main/java/ru/timmson/kanban/game/model/Column.java
package ru.timmson.kanban.game.model;

import lombok.Builder;
import lombok.Getter;
import lombok.Setter;

import java.util.LinkedList;
import java.util.List;

public class Column {

    @Getter
    private Stage stage;

    private LinkedList<Card> toDoQueue;

    private LinkedList<Card> wipQueue;

    @Getter
    private LinkedList<Card> doneQueue;

    @Getter
    @Setter
    private Integer wipLimit;

    @Getter
    private Boolean isOutputQueueLimited;

    @Builder
    Column(LinkedList<Card> toDoQueue, Stage stage, Integer wipLimit, Boolean isOutputQueueLimited) {
        this.stage = stage;
        this.toDoQueue = toDoQueue;
        this.wipQueue = new LinkedList<>();
        this.doneQueue = new LinkedList<>();
        this.wipLimit = wipLimit;
        this.isOutputQueueLimited = isOutputQueueLimited;
    }

    public Integer getWipSize() {
        return this.wipQueue.size();
    }

    public void handle(List<Dice> dices, Integer currentDay) {
        Integer points = dices.stream().mapToInt(dice -> dice.roll(stage)).sum();
        pull(currentDay);
        Card card = !wipQueue.isEmpty() ? wipQueue.getFirst() : null;
        while (card != null && points > 0) {
            points = card.decreaseRemaining(stage, points);
            if (card.getRemaining(stage) == 0) {
                if (stage.equals(Stage.STAGE_TESTING)) {
                    card.setEndDay(currentDay);
                }
                doneQueue.push(wipQueue.pop());
                pull(currentDay);
                card = !wipQueue.isEmpty() ? wipQueue.getFirst() : null;
            }
        }
        pull(currentDay);
    }

    private void pull(Integer currentDay) {
        while (wipLimit > (wipQueue.size() + (isOutputQueueLimited ? doneQueue.size() : 0)) && !toDoQueue.isEmpty()) {
            Card newCard = toDoQueue.pop();
            if (stage.equals(Stage.STAGE_ANALYSIS)) {
                newCard.setStartDay(currentDay);
            }
            wipQueue.push(newCard);
        }
    }
}

File: src/main/kotlin/ru/timmson/kanban/game/model/Column.kt
package ru.timmson.kanban.game.model

import java.util.*

class Column(var stage: Stage, var wipLimit: Int, private val toDoQueue: LinkedList<Card>, val isOutputQueueLimited: Boolean) {

    private val wipQueue = LinkedList<Card>()

    val doneQueue = LinkedList<Card>()

    fun getWipSize(): Int {
        return wipQueue.size
    }

    fun handle(dices: List<Dice>, currentDay: Int) {
        var points = dices.sumBy { dice -> dice.roll(stage)!! }
        pull(currentDay)
        var card = (if (!wipQueue.isEmpty()) wipQueue.first else null)
        while (card != null && points > 0) {
            points = card.decreaseRemaining(stage, points)!!
            if (card.getRemaining(stage) == 0) {
                if (stage == Stage.STAGE_TESTING) {
                    card.endDay = currentDay;
                }
                doneQueue.push(wipQueue.pop())
                pull(currentDay)
                card = (if (!wipQueue.isEmpty()) wipQueue.first else null)
            }
        }
    }

    fun pull(currentDay: Int) {
        while (wipLimit > wipQueue.size + (if (isOutputQueueLimited) doneQueue.size else 0) && !toDoQueue.isEmpty()) {
            val card = toDoQueue.pop()
            if (stage == Stage.STAGE_ANALYSIS) {
                card.startDay = currentDay
            }
            wipQueue.push(card)
        }
    }

}